### PR TITLE
Support multiple slot lengths

### LINKED ISSUE #282: Support multiple slot lengths
See `Ouroboros.Consensus.BlockchainTime`.

### PR DESCRIPTION
**Pull request by edsko**
The main change in this PR is the generalization of the real blockchain time from taking a single slot length to (a value derived from)

```haskell
data SlotLengths = SlotLengths {
      -- | Slot length in current segment
      currentSlotLength :: SlotLength

      -- | Segment length and future slot lengths (if hard fork planned)
    , nextSlotLengths   :: Maybe (SegmentLength, SlotLengths)
    }
```

recording all slot lengths across the chain. 

Testing story: I have generalized the already reasonably extensive property tests that we had for the original blockchain time infrastructure. In addition, I've also generalized the BFT dynamic tests so that it can have multiple slot lengths, and added a test label:

```
ouroboros-consensus
  Dynamic chain generation
    simple BFT convergence: OK (106.33s)
      +++ OK, passed 1000 tests.
      
      consensus expected (1000 in total):
      100.0% True
      
...
      
      slot length changes (1000 in total):
      95.3% 0
       4.7% 1
```

Closes #282.

### CODE REVIEW

File: ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/SlotLengths.hs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+            let startNext = segmentStartSlot (focusedStart next)
+                distance  = unSlotNo (startNext - startThis)
+            in conjoin [
+                   absSlot `lt` startNext
+                 , relSlot `lt` distance
+                 ]
+      ]
+  where
+    focused         = focusSlotLengths ls
+    startThis       = segmentStartSlot (focusedStart this)
```
Comment (edsko) [inline ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/SlotLengths.hs]: Reorder these two lines.

File: ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/SlotLengths.hs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+-------------------------------------------------------------------------------}
+
+infix 4 `lt`
+infix 4 `ge`
+
+lt :: (Show a, Ord a) => a -> a -> Property
+lt x y = counterexample (show x ++ " >= " ++ show y) $ x < y
+
+ge :: (Show a, Ord a) => a -> a -> Property
+ge x y = counterexample (show x ++ " < " ++ show y) $ x >= y
```
Comment (edsko) [inline ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/SlotLengths.hs]: Move to `Util` module?

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/Abstract.hs [hunk 3]
```diff
@@ -183,11 +181,8 @@ class ( Show (ChainState    p)
   -- | We require that protocols support a @k@ security parameter
   protocolSecurityParam :: NodeConfig p -> SecurityParam
 
-  -- | The slot length.
-  --
-  -- TODO: the 'SlotLength' will change after the hard fork, so this will need
-  -- to be modified.
-  protocolSlotLength :: NodeConfig p -> SlotLength
+  -- | The slot length (across all hard forks)
+  protocolSlotLength :: NodeConfig p -> SlotLengths
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/Abstract.hs]: Rename to `protocolSlotLengths`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/BFT.hs [hunk 4]
```diff
@@ -109,7 +110,7 @@ data BftParams = BftParams {
     , bftNumNodes      :: !Word64
 
       -- | Slot length
-    , bftSlotLength    :: !SlotLength
+    , bftSlotLength    :: !SlotLengths
```
Comment (edsko) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/BFT.hs]: Needs a comment explaining why.

File: ouroboros-consensus/test-consensus/Test/Dynamic/General.hs [hunk 5]
```diff
@@ -139,7 +140,6 @@ runTestNetwork pInfo
       nodeTopology
       pInfo
       (seedToChaCha seed)
-      slotLen
   where
     slotLen :: DiffTime
```
Comment (edsko) [inline ouroboros-consensus/test-consensus/Test/Dynamic/General.hs]: This should not define its own slot length, but rather use the one passed in.

File: ouroboros-consensus/tools/db-convert/Main.hs [hunk 6]
```diff
@@ -189,6 +192,8 @@ validateChainDb dbDir genesisConfig onlyImmDB verbose =
             putStrLn $ "DB tip: " ++ condense chainDbTipPoint
           )
   where
+    -- TODO: Is this right, can we just assume any odd slot length here?
+    -- (And if so, could we use fixedBlockchainTime instead?)
     slotLength = slotLengthFromMillisec (20 * 1000)
```
Comment (edsko) [inline ouroboros-consensus/tools/db-convert/Main.hs]: Justification: converting old chain, which indeed has 20s slot length.
Use `slotLengthFromSec.`

File: ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/WallClock.hs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+instance Arbitrary TestDelayIO where
+  arbitrary = do
+      tdioStart'  <- arbitrary
+      tdioSlotLen <- slotLengthFromMillisec <$> choose (1, 1_000)
+      return TestDelayIO{..}
+
+-- | Just as a sanity check, also run the tests in IO
+--
+-- We override the maxinum number of tests since there are slow.
```
Comment (edsko) [inline ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/WallClock.hs]: Add comment: this test _can_ fail if process under heavy load (the slot after the delay might be later than the one we expect).

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime/Mock.hs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      getCurrentSlot = return slot
+    , onSlotChange_  = const (return ())
+    }
+
+{-------------------------------------------------------------------------------
+  Testing time
+-------------------------------------------------------------------------------}
+
+-- | Number of slots
+newtype NumSlots = NumSlots Word64
```
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime/Mock.hs]: Happy to see `Int` changed to `Word64` here.

**Pull request merged**